Q1

In [1]:
import pandas as pd

# Load the dataset
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv'
df = pd.read_csv(url)

# Check for missing values
missing_values = df.isnull().sum()

# Display missing values per column
print(missing_values)

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64


Q2

In [26]:
import pandas as pd

# Load the dataset
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv'
df = pd.read_csv(url)

# Get the number of rows and columns
num_rows, num_columns = df.shape

num_rows, num_columns


(391, 11)

Q2.2

OBSERVATIONS: An observation is each entry of your given dataset. For example, each villager will be an observation. So there are 391 observations in the villagers dataset. Observations are the rows of a dataset. 

VARIABLES: Variables are different characteristics/attributes that are used for each observation (entry). Variables  are the columns of a dataset.

Q3

In [97]:
import pandas as pd

# Load the dataset
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv'
df = pd.read_csv(url)

# Get summary statistics for numerical columns
summary_statistics = df.describe()
summary_statistics

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

In [24]:
# Value counts for the 'species' column
species_counts = df['species'].value_counts()
print("Species Counts:")
species_counts

# Value counts for the 'gender' column
gender_counts = df['gender'].value_counts()

print("\nGender Counts:")
gender_counts

# Value counts for other columns as needed


Species Counts:
species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear         15
mouse        15
horse        15
bird         13
penguin      13
sheep        13
elephant     11
wolf         11
ostrich      10
deer         10
eagle         9
gorilla       9
chicken       9
koala         9
goat          8
hamster       8
kangaroo      8
monkey        8
anteater      7
hippo         7
tiger         7
alligator     7
lion          7
bull          6
rhino         6
cow           4
octopus       3
Name: count, dtype: int64

Gender Counts:


gender
male      204
female    187
Name: count, dtype: int64

Q4

In [38]:
import pandas as pd

# Load the dataset
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [111]:
import pandas as pd

url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [33]:
import pandas as pd

url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

df.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [92]:
import pandas as pd

url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

df.shape

(891, 15)

a) df.shape gives the total number of rows and columns. df.describe() gives summary statistics (count, mean, std, min, 25%, 50%, 75%, and max) for each variable. Since this dataset includes non-numeric variables, it is inevitable that df.describe() will not use non-numeric variables to calculate numeric statistics. It automatically removes any missing values.

b) df.shape counts the total number of observations (ie. the rows of this dataset) even if there is missing data. When we use df.describe(), the count for certain variables (columns) may be fewer as a result of missing data. For example, the variable "age" has a lower count. This is a direct result of there being fewer observations with data for the "age" variable. df.describe() automatically removes any missing values when calculating the statistics for each given variable with numeric data.

Q5

Attribute: An attribute is a value of an object. It outputs characteristics of the object. This means it will not take any arguments (inputs). Thus it does not end with parentheses.

Method: Methods, on the other hand, process information that is given, or not given, through the arguments (input). They need brackets (even if it is empty) to execute the code in the method. Even if the parentheses are empty, it will take the variable before the '.' as the argument (e.g. "df" in "df.describe()"). This is why methods require parentheses.

This is my chat log for STA130 Week 01 Prelecture Homework: https://chatgpt.com/share/fd6577b9-91a5-4a4b-8a03-9cb8293bca5b

Q6

count: Number of valid (not missing) data in a variable (column). It shows the number of observations for each variable respectively.

mean: This is the average value of a numeric variable (column). This is calculated by taking the sum of all observations and then dividing it by the total number of observations for a given variable.

std: This stands for "standard deviation." This basically quantifies the dispersion of data in a variable. Greater number for standard deviation means there is a greater spread in numerical data.

min: Min is the minimum value in the data of a variable (column).

25%: This is the data at the the 25th percentile (1st quartile) of a numeric variable when the data is in ascending order. 

50%: This is also known as the median. It is the middle value of a numeric variable when the data is in ascending order.

75%:This is the data at the the 75th percentile (3rd quartile) of a numeric variable when the data is in ascending order. 

max: This is the greatest value in the data of a variable (column)

Chat log for Q6: https://chatgpt.com/share/9474bf2e-b111-41bf-8e13-1061ebe30860

Q7 (chatbot not used for this question as I was able to think of examples from the lecture)

1. Provide an example of a "use case" in which using df.dropna() might be peferred over using del df['col']

An example use case would be when you want to change a data type for a column. You may want to drop all missing values to be able to convert, for say, all variables with a type float64 to an int64. If there are missing values, the computer will not be able to convert non-finite values (ie NA).  del df['col'] will be tedious to do this process and there may be important variables that cannot get lost.

2. Provide an example of "the opposite use case" in which using del df['col'] might be preferred over using df.dropna()

If one does not need a certain variable (maybe it ends up being useless), they could use del df['col'] to remove a whole column. Using df.dropna() will be useless as it will remove all entries with missing values rather than a column as a whole .

3. Discuss why applying del df['col'] before df.dropna() when both are used together could be important

If you use df.dropna() first, you will be removing ALL observations with missing values. This means if there is a variable named 'col', then the observations with missing values for that variable will be removed when executing df.dropna(). This means that more observations will be removed than neccessary since the 'col' was going to be deleted anyways. Deleting the column first will make it so that the observations with missing data for 'col' will not be removed unless they have missing data for other variables.

4. Remove all missing data from one of the datasets you're considering using some combination of del df['col'] and/or df.dropna() and give a justification for your approach, including a "before and after" report of the results of your approach for your dataset.

I would personally use df.dropna() to remove all missing data. This would be more efficient as it is fewer lines of code than going through each column. Also, removing an entire column will remove an entire variable; maybe certain variables are still useful, so it may be a bad idea to drop the entire column. This means that useful data may be lost. Thus, the approach of simply using df.drop() will be more effective.

In [10]:
import pandas as pd

url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

# Before
df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [9]:
import pandas as pd

url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

# After
df = df.dropna()
df.isna().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
deck           0
embark_town    0
alive          0
alone          0
dtype: int64

Q8

Q8.1

Using the titanic dataset, lets say you would like to see the statistics for the survivors and non-survivors.
In this case, you could use "groupby('col1')" method to group the data within a column/variable.
Using this, we can access the ages from that grouped variable using ['col2'] and use "describe()" to give us statistics about the age of the survivors/non-survi ors.

ChatGPT interaction: https://chatgpt.com/share/5bd59fef-0533-4c5d-a6d1-736b39cae380

In [25]:
import pandas as pd

url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

df.groupby('survived')["age"].describe()

count       mean        std   min   25%   50%   75%   max
survived                                                           
0         424.0  30.626179  14.172110  1.00  21.0  28.0  39.0  74.0
1         290.0  28.343690  14.950952  0.42  19.0  28.0  36.0  80.0

In [26]:
df.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

Q8.2

df.describe() will count the total entries (excluding missing values) of a column. It will only count valid data.
df.groupby('col1')["col2"].describe() has different values due many reasons.

Firstly, the groupby('col1') will only take a certain column and categorize it by each entry. For example, 'survived' column will be grouped into either 0 or 1 (True or False). So, for each group, it will look like there are a lower count. However, the sum will be the same as shown in df.describe() for the given column.

Additionally, we are also accessing a secondary column. This secondary column may have missing values. If it does, then the observation a missing value in that secondary column will be removed (ie if an entry has a value for col1 but not col2, it will be removed automatically). 

Keeping these in mind, it is clear that the count shown using df.groupby('col1')["col2"].describe() will look a lot lower than the count for the columns using df.describe(). Thus df.describe() will capture the total count of non-missing data for each variable, but df.groupby('col1')["col2"].describe() will capture the count of a specific group of a category that also has a value in a secondary variable.

Q8.3

A) ChatGPT is better: ChatGPT is definitely a faster in solving this problem compared to google. Google, you have to search, enter a website, and scroll while you can get the answer straight up through ChatGPT.

B) ChatGPT is better: ChatGPT solves this problem quickly. Google, again, you would have to look it up which may take longer.

C) ChatGPT is better: ChatGPT is able to understand that I meant to use df using previous context. Google does not know this, so it brought me to various questions that are related pandas, but not specifically my situation.

D) ChatGPT is better: ChatGPT knew the exact error and told me that I am missing parentheses. Google brought me to many pages related to inputs and syntax errors which was mostly useless.

E) ChatGPT is better: ChatGPT tells me the correct name of the method. It recognizes the correct spelling. Google brought me to a stackoverflow page with a different error related to 'pct'(?).

F) ChatGPT is better: Although ChatGPT brought me a number of solutions, it told me the general idea. It did a good job analyzing what could have caused it. ChatGPT gave me some code that would check if the columns even exist in the first place which is definitely useful. Google, again, was not that useful as most "KeyErrors" were unrelated to my specific problem.

G) ChatGPT is better: By giving it some context about the function I was trying to use, it told me that I must have quotes around a column name to avoid the NameError. This was the same for missing the quotes for col2. Google seemed to bring me to other people's code who had the same error. The only problem was that it may have been hard to navigate where and what the error is since their code was larger.

Here is my interaction with ChatGPT: https://chatgpt.com/share/66e34f52-ab5c-8008-a18e-c0353df6e305

In [1]:
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

NameError: name 'pd' is not defined

In [2]:
import pandas as pd

url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv'
df = pd.read_csv(url)

HTTPError: HTTP Error 404: Not Found

In [4]:
import pandas as pd
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

DF.groupby("pclass")["age"].describe()

NameError: name 'DF' is not defined

In [5]:
import pandas as pd
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url

SyntaxError: incomplete input (887052805.py, line 3)

In [6]:
import pandas as pd
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

df.groupby('pclass')['age'].describle()

AttributeError: 'SeriesGroupBy' object has no attribute 'describle'

In [9]:
import pandas as pd
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

df.groupby("sex")["Age"].describe()

KeyError: 'Column not found: Age'

In [11]:
import pandas as pd
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv'
df = pd.read_csv(url)

df.groupby("sex")[age].describe()

NameError: name 'age' is not defined

Q9
Yes